src/goto-instrument/contracts/assigns.cpp

```diff
@@ -9,144 +9,179 @@ Date: July 2021
+
+  const auto validity_check_expr =
+    and_exprt{all_dereferences_are_valid(source_expr, parent.ns),
+              w_ok_exprt{slice.first, slice.second}};
+  snapshot_instructions.add(goto_programt::make_assignment(
+    validity_condition_var, validity_check_expr, location));
+
+  snapshot_instructions.add(goto_programt::make_goto(
+    skip_target, not_exprt{validity_condition_var}, location));
+
+  snapshot_instructions.add(goto_programt::make_assignment(
+    lower_bound_address_var, slice.first, location));
+
+  snapshot_instructions.add(goto_programt::make_assignment(
+    upper_bound_address_var,
+    minus_exprt{plus_exprt{slice.first, slice.second},
+                from_integer(1, slice.second.type())},
+    location));
+
+  snapshot_instructions.destructive_append(skip_program);
+
+  return snapshot_instructions;
 }
 
-exprt assigns_clauset::targett::normalize(const exprt &expr)
+const slicet assigns_clauset::conditional_address_ranget::normalize_to_slice(
+  const exprt &expr) const
 {
-  if(expr.id() != ID_address_of)
-    return expr;
+  // FIXME: Refactor these checks out to a common function that can be
+  // used both in compilation and instrumentation stages
+  if(can_cast_expr<side_effect_expr_function_callt>(expr))
+  {
+    const auto &arg =
+      get_cprover_object_call_argument(to_side_effect_expr_function_call(expr));
+    return {
+      minus_exprt{
+        typecast_exprt::conditional_cast(arg, pointer_type(char_type())),
+        pointer_offset(arg)},
+      typecast_exprt::conditional_cast(object_size(arg), signed_size_type())};
+  }
+  else if(is_lvalue(expr))
+  {
+    const auto &size = size_of_expr(expr.type(), parent.ns);
+
+    INVARIANT(
+      size.has_value(),
+      "`sizeof` must always be computable on an assigns clause target.");
 
-  const auto &object = to_address_of_expr(expr).object();
-  if(object.id() != ID_index)
-    return object;
+    return {typecast_exprt::conditional_cast(
+              address_of_exprt{expr}, pointer_type(char_type())),
+            typecast_exprt::conditional_cast(size.value(), signed_size_type())};
+  }
 
-  return to_index_expr(object).array();
+  UNREACHABLE;
+}
+
+exprt assigns_clauset::conditional_address_ranget::generate_inclusion_check(
+  const conditional_address_ranget &lhs,
+  bool skip_lhs_validity_check) const
+{
+  const auto &inclusion_check = conjunction(
+    {validity_condition_var,
+     same_object(lower_bound_address_var, lhs.lower_bound_address_var),
+     same_object(lhs.upper_bound_address_var, upper_bound_address_var),
+     less_than_or_equal_exprt{lower_bound_address_var,
+                              lhs.lower_bound_address_var},
+     less_than_or_equal_exprt{lhs.upper_bound_address_var,
+                              upper_bound_address_var}});
+
+  if(skip_lhs_validity_check)
+    return inclusion_check;
+
+  return or_exprt{not_exprt{lhs.validity_condition_var}, inclusion_check};
```

Comment:
Is this suppose to be unreachable?

Reply:
Yes, removed dead code and made it private.